src/disk_io_thread.cpp

```diff
@@ -261,16 +261,23 @@ constexpr disk_job_flags_t disk_interface::cache_hit;
 
 		TORRENT_ASSERT(m_magic == 0x1337);
 		m_magic = 0xdead;
+		TORRENT_ASSERT(m_generic_io_jobs.m_queued_jobs.empty());
+		TORRENT_ASSERT(m_hash_io_jobs.m_queued_jobs.empty());
 	}
 #endif
 
 	void disk_io_thread::abort(bool const wait)
 	{
+		// first make sure queued jobs have been submitted
+		// otherwise the queue may not get processed
+		submit_jobs();
+
 		// abuse the job mutex to make setting m_abort and checking the thread count atomic
 		// see also the comment in thread_fun
 		std::unique_lock<std::mutex> l(m_job_mutex);
 		if (m_abort.exchange(true)) return;
-		bool const no_threads = m_num_running_threads == 0;
+		bool const no_threads = m_generic_threads.num_threads() == 0
+			&& m_hash_threads.num_threads() == 0;
```

Comment:
I think the assert in CI is caused by this being `true`, even though there are threads that have been spawned, they just don't "count" yet. Leading to the last thread also calling `abort_jobs()`, just as is done a few lines below here. And there seems to be an assert that it's only called once.

Reply:
When threads are spawned they are added to the thread vector while holding `m_job_mutex` so I don't think it could be the startup path. I think the problem is the thread shutdown path. The last thread calls `abort_jobs` after it has decremented `m_num_running_threads` and released `m_job_mutex`, so I don't see any easy way to detect a thread which might be lurking at the end of `thread_fun` right before the call to `abort_jobs`. We may need to remove the assert and rely on `m_jobs_aborted` to prevent a double invocation.